src/libsyntax/parse/parser.rs

```diff
@@ -2700,6 +2697,99 @@ impl<'a> Parser<'a> {
         self.maybe_recover_from_bad_qpath(expr, true)
     }
 
+    fn parse_async_macro_or_stmt(
+        &mut self,
+        lo: Span,
+        await_sp: Span,
+    ) -> PResult<'a, (Span, ExprKind)> {
+        Ok(match self.token {
+            token::Not => {
+                // Handle correct `await!(<expr>)`
+                // FIXME: make this an error when `await!` is no longer supported
+                // https://github.com/rust-lang/rust/issues/60610
+                self.expect(&token::Not)?;
+                self.expect(&token::OpenDelim(token::Paren))?;
+                let expr = self.parse_expr().map_err(|mut err| {
+                    err.span_label(
+                        await_sp,
+                        "while parsing this await macro call",
+                    );
+                    err
+                })?;
+                self.expect(&token::CloseDelim(token::Paren))?;
+                (expr.span, ExprKind::Await(ast::AwaitOrigin::MacroLike, expr))
+            }
+            token::Question => {
+                // Handle `await? <expr>`
+                self.bump(); // `?`
+                let expr = self.parse_expr().map_err(|mut err| {
+                    err.span_label(
+                        await_sp,
+                        "while parsing this incorrect await statement",
+                    );
+                    err
+                })?;
+                let sp = lo.to(expr.span);
+                let expr_str = self.sess.source_map().span_to_snippet(expr.span)
+                    .unwrap_or_else(|_| pprust::expr_to_string(&expr));
+                let expr = self.mk_expr(
+                    sp,
+                    ExprKind::Await(ast::AwaitOrigin::FieldLike, expr),
+                    ThinVec::new(),
+                );
+                let mut err = self.struct_span_err(
+                    sp,
+                    "incorrect use of `await`",
+                );
+                err.span_suggestion(
```

Comment:
Seems preferable to chain this (+ `.emit()`) ? (unless you deduplicate the `span_suggestion`)